pennylane/ops/qubit/matrix_ops.py

```diff
@@ -332,3 +334,134 @@ def _controlled(self, control):
 
     def label(self, decimals=None, base_label=None, cache=None):
         return super().label(decimals=decimals, base_label=base_label or "U", cache=cache)
+
+
+class BlockEncode(Operation):
+    r"""BlockEncode(a, wires)
+    Apply an arbitrary matrix, :math:`A`, encoded in the top left block of a unitary matrix.
+
+    .. math::
+
+        \begin{align}
+             U(A) &=
+             \begin{bmatrix}
+                A & \sqrt{I-AA^\dagger} \\
+                \sqrt{I-A^\dagger A} & -A^\dagger
+            \end{bmatrix}.
+        \end{align}
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 1
+    * Number of dimensions per parameter: (2,)
+    * Gradient recipe: None
+
+    Args:
+        A (array[complex]): a general :math:`(n\times m)`matrix to be encoded
+        wires (Sequence[int] or int): the wire(s) the operation acts on
+        do_queue (bool): indicates whether the operator should be
+            recorded when created in a tape context
+        id (str): custom label given to an operator instance,
+            can be useful for some applications where the instance has to be identified
+
+    Raises:
+        ValueError: if the number of wires doesn't fit the dimensions of the matrix
```

Comment:
Maybe we discussed before, but how many qubits do you need for this op if `A` is `(3, 3)`?

Reply:
If A is (3, 3), we need 3 qubits.
In general, for A that is (n, m), we need ceil(log2(n+m)) qubits.

Reply:
Yup, exactly! It's padded with identity in the bottom right and zeros elsewhere.